### PR TITLE
feat: puya debugging support

### PR DESCRIPTION
**Pull request by aorumbayev**
# Proposed changes 

- Introduces support for puya sourcemap debugging while maintaining baseline set of features compatible with TEAL
- Adds minor improvements to launch configs, allowing starting a session by passing the contents of sources.json directly in launch config. 
- Additional test and set of extra sample workspaces for Puya (showcasing scenarios with pure Puya (with/without) third party source references and with an intermix of Puya and Teal sourcemaps
- Minor documentation for sample workspaces folder
- Tweaking the sample extension to allow 'py' files for setting breakpoints. Note that the actual extension published under title 'AlgoKit AVM VScode Debugger' is published from a separate repo that depends on this repository's content - the DAP package 

## TODO: 

- [x] - Address backward stepping bug

### CODE REVIEW

**Comment by aorumbayev** [comment]
@daniel-makerx pushed a fix for a curious edge case found between SimulateResponse objects generated via get_obj_for_encoding in v2 which is deprecated in v3. V3 beta was already used on the base of this repo so i bumped it to final v3 release, and added a method for converting the v2 version of simulate response to v3. Given that algokit extension is already relying on avm-debbugger with beta v3 algosdk -> this might already be affecting python templates with prod presets and typescript deployers (which will opt in to storing debug traces with v6 utils that relies on < v3.0 algosdk versions which will persist sourcemaps in a format that is not compatible with v3's format when loading json encoded SimulateResponse). Introducing this fix would allow pre v3 algosdk/v7-v8 utils-ts users to use this extension without upgrading anything else. 

Also added some more tweaks to unit tests and added a ui tweak to hide the Locals when its not puya frontend frames.

Ready for review

File: extension/src/configuration.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ export class AvmDebugConfigProvider
   implements vscode.DebugConfigurationProvider
 {
   /**
-   * Massage a debug configuration just before a debug session is being launched,
+   * Message a debug configuration just before a debug session is being launched,
```
Comment (neilcampbell) [inline extension/src/configuration.ts]:
```suggestion
   * Massage a debug configuration just before a debug session is being launched,
```

I think massage is the correct term here.

File: sampleWorkspace/README.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+## How to navigate examples?
+
+This folder contains sample workspaces representing various debugger use cases:
+
+- Simulate traces, TEAL sources and sourcemaps (e.g. `stack-scratch`, `stepping-test`, `app-state-changes`, `errors`). Various examples of using the debugger with TEAL sourcemaps and sources.
+- Simulate traces, [Puya](https://github.com/algorandfoundation/puya) sources and sourcemaps (e.g. `puya`, `puya_third_parties`). Showcasing various ways to simulate traces with puya sourcemaps and sources.
```
Comment (neilcampbell) [inline sampleWorkspace/README.md]:
```suggestion
- Simulate traces, [Puya](https://github.com/algorandfoundation/puya) sources and sourcemaps (e.g. `puya`). Showcasing various ways to simulate traces with puya sourcemaps and sources.
```

File: sampleWorkspace/README.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+- Simulate traces, TEAL sources & sourcemaps and Puya sources & sourcemaps (e.g. `puya_and_teal`). Showcasing the ability to mix and match TEAL and Puya frontend languages within the same debug session.
```
Comment (neilcampbell) [inline sampleWorkspace/README.md]:
```suggestion
```

File: src/common/debugSession.ts [hunk 4]
```diff
@@ -327,57 +362,51 @@ export class AvmDebugSession extends DebugSession {
...
-
-      // The runtime has a stack where the latest call is the last element. We need to return the
-      // reverse of that.
-      const adjustedEndFrame = this._runtime.stackLength() - startFrame;
-      const adjustedStartFrame = Math.max(0, adjustedEndFrame - maxLevels);
+      const startFrame = args.startFrame || 0;
+      const maxLevels = args.levels || 1000;
 
-      const stk = this._runtime.stack(adjustedStartFrame, adjustedEndFrame);
+      const stk = this._runtime.stack(startFrame, startFrame + maxLevels);
```
Comment (neilcampbell) [inline src/common/debugSession.ts:368]: Assuming this was done on purpose?

**Review by neilcampbell — APPROVED** [review]

File: src/common/debugSession.ts (hunk 4 above)
Comment (daniel-makerx) [reply src/common/debugSession.ts:368]: Yep the stack used to be in reverse order, so the parameters used to be adjusted to query with inverted indexes. That is no longer required

Comment (daniel-makerx) [reply src/common/debugSession.ts:368]: See https://github.com/algorandfoundation/avm-debugger/pull/2/files/e5546a104221e452aa4a4916665609efb04a6a28#diff-57714ab2de4315a6d3e2e1e17241bd601aabb7f0c09bd26debf698925f8493d4R224

**Pull request merged**